### ISSUE TITLE
foreach XmlNodeList the XmlNode from SelectSingleNode is not allocating correctly

### ISSUE THREAD

**Issue by jakobkellett** [Microsoft/AL#4723]
### **Describe the bug**
**I am calling a web service that returns the following XML:**

```
<content>
    <read refname="accountquery" responsestatus="success">
       <activeaccounts>
         <id>ae4e36fe-ff19-4587-96ea-cdb7770efa32</id>
         <name>First Credit Account</name>
         <platformtype.name>Credit Card</platformtype.name>
       </activeaccounts>
       <activeaccounts>
         <id>75a0218f-e111-4b4a-9689-db1b5186b3b6</id>
         <name>Second Credit Account</name>
         <platformtype.name>Credit Card</platformtype.name>
       </activeaccounts>
    </read>
</content>
```

**I have the following local variables:**
```
CurrentMessage: Text;
LocalXmlDocument: XmlDocument;
LocalNode: XmlNode;
LocalValueNode: XmlNode;
LocalXmlElement: XmlElement;
LocalNodeList: XmlNodeList;
```

What is happening is the first iteration processes just fine, but on the second iteration of the foreach is acting strange.
The LocalNode appears to allocate correctly, but the LocalNode.SelectSingleNode('//id', LocalValueNode) is grabbing the previous value of LocalNode for some reason.

**To Reproduce**
Steps and/or AL code to reproduce the behavior:
 ```
 LocalXmlDocument.SelectNodes('//read[@refname="accountquery"]//activeaccounts', LocalNodeList);
   if LocalNodeList.Count() > 0
            then begin
                foreach LocalNode in LocalNodeList do begin
                    LocalNode.WriteTo(CurrentMessage);
                    Message(CurrentMessage);

                    LocalNode.SelectSingleNode('//id', LocalValueNode);
                    LocalXmlElement := LocalValueNode.AsXmlElement();
                    LocalValueNode.WriteTo(CurrentMessage);
                    Message(CurrentMessage);
                    Message(LocalXmlElement.InnerText());
                end;
   end else begin
          Message('There were no accounts setup.')
   end;
```

**Expected behavior**
The iteration should be setting LocalValueNode (XmlNode) to 75a0218f-e111-4b4a-9689-db1b5186b3b6 on the second iteration of the foreach LocalNodeList (XmlNodeList).

**Screenshots**
----
Iteration 1 - LocalNode WriteTo Text
![iteration-1-LocalNode](https://user-images.githubusercontent.com/1669386/54332636-4c7ea280-45dc-11e9-83d1-c02da8ecba75.png)
----
Iteration 1 - LocalValueNode WriteTo Text
![iteration-1-LocalValueNode](https://user-images.githubusercontent.com/1669386/54332637-4f799300-45dc-11e9-8ebd-6af9d3c91373.png)
----
Iteration 1 - LocalXmlElement InnerText
![iteration-1-LocalXmlElement](https://user-images.githubusercontent.com/1669386/54332639-51435680-45dc-11e9-9007-bf4168c135aa.png)
----
Iteration 2 - LocalNode WriteTo Text
![iteration-2-LocalNode](https://user-images.githubusercontent.com/1669386/54332641-530d1a00-45dc-11e9-9f31-aac9cbd6f958.png)
----
Iteration 2 - LocalValueNode WriteTo Text
![iteration-2-LocalValueNode](https://user-images.githubusercontent.com/1669386/54332644-56080a80-45dc-11e9-91df-35ea1f2c1bbb.png)
----
Iteration 2 - LocalXmlElement InnerText
![iteration-2-LocalXmlElement](https://user-images.githubusercontent.com/1669386/54332648-57d1ce00-45dc-11e9-93d1-13739b8e4a14.png)
----

**Versions:**

 - AL Language: 2.1.79379

 - Business Central: US Dynamics NAV 13.3 (29060)

**Comment by jakobkellett** [comment]
Hey @kalberes , I appreciate the labeling of the bug. Is there any updates by chance on this issue? We will be going to the Microsoft Directions Event in Early May (bought a booth!) and hoping we can convert our code from C/AL in time! Thank you~

**Comment by xgaronnat** [comment]
Hi,
I found the same issue with this code in April '19 CU1 : Line is always giving the value of the first node(Line) only...


```
XmlDoc.SelectNodes('//ShipmentItem', Lines);
        if Lines.Count > 0 then begin
            message('%1 lines found', Lines.Count);
            foreach Line in Lines do begin

                Line.SelectSingleNode('//ItemID/ID', XmlSubNode);
                ItemID := XmlSubNode.AsXmlElement().InnerText;

                Line.SelectSingleNode('//ShippedQuantity', XmlSubNode);
                ShippedQty := XmlSubNode.AsXmlElement().InnerText;

                Line.SelectSingleNode('//LineNumber', XmlSubNode);
                LineNumber := XmlSubNode.AsXmlElement().InnerText;

                message('Line\Item ID = %1\Shipped Qty = %2\Line Number = %3',
                    ItemID,
                    ShippedQty,
                    LineNumber)
            end;
        end;
`

```

**Comment by xgaronnat** [comment]
Indeed it's working : use ./ItemID/ID for find the node in the current Line.

**Comment by ahelwig** [comment]
Replacing the "/" or "//" at the start of each XPath string with the context item expression "." followed by the "/" or "//" did the trick. Glad I found your post.

**Comment by MortenRa** [comment]
@atoader Is this going to be looked into

**Comment by atoader** [comment]
Yes, but I cannot provide an ETA. We have a large number of issues reported through GitHub.  We will work on prioritizing issues based on severity and input from developers in the next couple of weeks. We should have an update soon.

**Comment by SBalslev** [comment]
Hi! Thank you for this detailed report about XmlNodeList iteration and SelectSingleNode behavior.

This is a runtime/platform API issue rather than an AL compiler or developer tools issue. The XML API behavior is handled by the Business Central server runtime, which is outside the scope of this repository.

We're closing this to keep the backlog focused on AL tooling. If this is still an issue, we recommend posting it on https://aka.ms/BCIdeas or opening a support case for the runtime team to investigate.

Thank you for the detailed repro steps — they're very helpful for anyone investigating this further!
